core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,79 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.syntax.group._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+class Perm private(val mapping: Map[Int, Int]) extends (Int => Int) {
+  assert(mapping.forall(Function.tupled(_ != _)), "Perm contains a redundant mapping.")
+  require(mapping.values.toSet == mapping.keySet, "Image and preimage must be the same.")
+
+  override def apply(k: Int): Int = mapping.getOrElse(k, k)
+
+  override def toString: String = {
+    mapping.toSeq.sorted
```

Comment:
Provided this `.toString` is worthwhile in the first place, we might want to use `SortedMap` rather than sorting in `.toString`, or forget about sorting altogether (doesn't really with performance anywhere; this is somewhat cosmetic).

Reply:
I say this is good enough.